blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,866 @@
+    /// self.
+    pub fn pending(&self) -> (Vec<SignedMessage>, Tipset) {
+        let mut out: Vec<SignedMessage> = Vec::new();
+        for (addr, _) in self.pending.clone() {
+            out.append(self.pending_for(&addr).unwrap().as_mut())
+        }
+        (out, self.cur_tipset.clone())
+    }
+
+    /// Return a Vector of signed messages for a given from address. This vector will be sorted by
+    /// each messsage's sequence. If no corresponding messages found, return None result type
+    fn pending_for(&self, a: &Address) -> Option<Vec<SignedMessage>> {
+        let mset = self.pending.get(a);
+        match mset {
+            Some(msgset) => {
+                if msgset.msgs.is_empty() {
+                    return None;
+                }
+
+                let mut msg_vec = Vec::new();
+
+                for (_, item) in msgset.msgs.clone() {
+                    msg_vec.push(item);
+                }
+
+                msg_vec.sort_by_key(|value| value.message().sequence());
+
+                Some(msg_vec)
+            }
+            None => None,
+        }
+    }
+
+    /// Return Vector of signed messages given a block header for self
+    pub fn messages_for_blocks(
+        &mut self,
+        blks: Vec<BlockHeader>,
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let mut msg_vec: Vec<SignedMessage> = Vec::new();
+        for block in blks {
+            let (umsg, mut smsgs) = self.api.messages_for_block(&block)?;
+            msg_vec.append(smsgs.as_mut());
+            for msg in umsg {
+                let smsg = self.recover_sig(msg)?;
+                msg_vec.push(smsg)
+            }
+        }
+        Ok(msg_vec)
+    }
+
+    /// Attempt to get the signed message given an unsigned message in message pool
+    pub fn recover_sig(&mut self, msg: UnsignedMessage) -> Result<SignedMessage, Error> {
+        let val = self
+            .bls_sig_cache
+            .get(&msg.cid().map_err(|err| Error::Other(err.to_string()))?)
+            .unwrap();
+        Ok(SignedMessage::new_from_fields(msg, val.clone()))
+    }
+
+    /// Return gas price estimate this has been translated from lotus, a more smart implementation will
+    /// most likely need to be implemented
+    pub fn estimate_gas_price(&self, nblocksincl: u64) -> Result<BigInt, Error> {
+        // TODO possibly come up with a smarter way to estimate the gas price
+        let min_gas_price = 0;
+        match nblocksincl {
+            0 => Ok(BigInt::from(min_gas_price + 2)),
+            1 => Ok(BigInt::from(min_gas_price + 1)),
+            _ => Ok(BigInt::from(min_gas_price)),
+        }
+    }
+
+    /// Load local messages into pending. As of  right now messages are not deleted from self's
+    /// local_message field, possibly implement this in the future?
+    pub fn load_local(&mut self) -> Result<(), Error> {
+        for (key, value) in self.local_msgs.clone() {
```

Comment:
shouldn't need to clone if you are only using references, just turn into an iterator

Reply:
the clone was used because in the case where calling the add function with value fails, local_msgs needs to remove that kv pair.

Reply:
yeah, but you're taking the reference of the key? Why would you need the owned value? Even if you did, should only clone when and what you need to, and not for every key and value

Reply:
But by removing the clone and just using references, issues arose where multiple borrows due to the deletion of an element of the hashmap that was being iterated over.

Reply:
yeah cloning the whole hashmap is really expensive, just create a vector of keys to remove, and after this loop remove all those keys from the hashmap.

Reply:
after a little bit of train and error, realized that self.local_msgs does not need to be a hashmap, changed it over to a vec in latest commit